File: tests/fixtures/api-frameworks/spring-webflux-functional/src/main/java/com/acme/books/BookHandler.java
package com.acme.books;

import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Component
public class BookHandler {

    public Mono<ServerResponse> list(ServerRequest request) {
        return ServerResponse.ok().body(Flux.empty(), Book.class);
    }

    public Mono<ServerResponse> get(ServerRequest request) {
        String isbn = request.pathVariable("isbn");
        return find(isbn)
                .flatMap(book -> ServerResponse.ok().body(Mono.just(book), Book.class))
                .switchIfEmpty(ServerResponse.notFound().build());
    }

    public Mono<ServerResponse> create(ServerRequest request) {
        return request.bodyToMono(NewBook.class)
                .flatMap(book -> ServerResponse.status(HttpStatus.CREATED).bodyValue(book));
    }

    private Mono<Book> find(String isbn) {
        return Mono.empty();
    }
}

File: tests/fixtures/kotlin/webflux-corouter/src/main/kotlin/com/acme/catalog/BookHandler.kt
package com.acme.catalog

import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.Positive
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.server.ServerRequest
import org.springframework.web.reactive.function.server.ServerResponse
import org.springframework.web.reactive.function.server.awaitBody
import org.springframework.web.reactive.function.server.bodyValueAndAwait
import org.springframework.web.reactive.function.server.buildAndAwait
import org.springframework.web.server.ResponseStatusException

/** Handles the catalog's book routes. */
@Component
class BookHandler(private val books: BookStore) {

    suspend fun all(request: ServerRequest): ServerResponse {
        val author = request.queryParamOrNull("author")
        return ServerResponse.ok().bodyValueAndAwait(books.all(author))
    }

    suspend fun byId(request: ServerRequest): ServerResponse {
        val id = request.pathVariable("id")
        val book = books.byId(id) ?: throw ResponseStatusException(HttpStatus.NOT_FOUND, "no such book")
        return ServerResponse.ok().bodyValueAndAwait(book)
    }

    suspend fun create(request: ServerRequest): ServerResponse {
        val body = request.awaitBody<NewBook>()
        val saved = books.add(body)
        return ServerResponse.status(HttpStatus.CREATED).bodyValueAndAwait(saved)
    }
}

/** A book as the catalog accepts it. */
data class NewBook(
    @field:NotBlank val title: String,
    @field:NotBlank val author: String,
    @field:Positive val pages: Int = 1,
)
